add new option: aya-case-fold

gilbertw1 opened the pull request:
This pull request adds support for the option `aya-case-fold`. This option essentially allows auto-yasnippet to create templates where variables that are identical apart from the case of the first character are linked together to a single variable with case in the resulting snippet is preserved. 

I know that sounds like a mouthful, but the easiest way to see the behavior is in this screencast (first with this setting disabled, then again with it enabled):

![screencast](http://drop.bryan.sh/GLVxyzsBkQ.gif)

I modeled this feature after the ability to find and replace text while preserving the case of the initial character: https://www.emacswiki.org/emacs/CaseFoldSearch.

The code is fairly straightforward. I'm now returning an alist from `aya--parse` that contains the `id` of the match, it's `value`, and `ucase` (whether or not the value was uppercase or not). That alist is used to both recreate the current line with proper casing and create a yas template that conditionally uppercases the value to match the original casing.

abo-abo commented on README.md:
````diff
@@ -151,6 +152,28 @@ Here's a yasnippet that makes use of `aya-tab-position`. You need to call
+count_of_~red = get_total("~Red");
+```
+
+Then calling `yas-create`, then `yas-expand`, and finally typing `blue`, the result
````
Should be `aya-create` and `aya-expand`. Also, I can't get this use case to work. Using yasnippet 20170310.1724.

gilbertw1 replied:
I've corrected the function names. I'm testing using the same version yasnippet. Quick question, are you updating the variable (`(setq aya-case-fold t)`) prior to calling `aya-create` to create the snippet?

abo-abo replied:
Oops, sorry; I've had a long day. Your example does work.

abo-abo commented:
Comments:

1. When does it make sense for `aya-case-fold` to be off? Seems like a straight upgrade.
2. Please use `cl-lib` in place of seq functions (`seq-some`, `seq-filter`) etc.
3. I'm not sure about when `alist-get` was introduced, please check.

gilbertw1 commented:
Hey,

1. I wasn't sure if this would be desired behavior for everyone. I could either A) make it on by default or B) remove the setting and just make this how it works all the time.
2. Will do.
3. Looks like [Emacs 25](https://github.com/emacs-mirror/emacs/blob/master/etc/NEWS.25). What's your target minimum version? I can use an alternative if it's <25.

Thanks!

gilbertw1 commented:
I've replaced the `seq-*` functions with `cl-*` functions, and I've replaced the `alist-get` function calls as well. Let me know how you' like me to handle 1.

abo-abo commented:
> make it on by default

Let's do it like that.

> What's your target minimum version?

24.3.

gilbertw1 commented:
Cool, I've defaulted the feature to on by default.

abo-abo commented:
Thanks.
